src/netlink.c

```diff
@@ -0,0 +1,362 @@
+	int flags = luaL_optinteger(L,5,0);
+
+	struct sockaddr_nl addr;
+	struct nlmsghdr *nlh;
+
+	memset(&addr,0,sizeof(addr));
+	addr.nl_pid=dstpid;
+	addr.nl_family=AF_NETLINK;
+	addr.nl_groups=groups;
+
+	nlh = (struct nlmsghdr *)malloc(NLMSG_SPACE(payload_size));
+	memset(nlh, 0, NLMSG_SPACE(payload_size));
+	nlh->nlmsg_len = NLMSG_SPACE(payload_size);
+	nlh->nlmsg_pid = nl->srcpid;
+	nlh->nlmsg_flags = flags;
+
+	memcpy(NLMSG_DATA(nlh),payload,payload_size);
+
+	int err = socket_sendto(&nl->fd,(char *)nlh,nlh->nlmsg_len,&sent,(struct sockaddr *)&addr,sizeof(addr),tm);
+	if(err!=IO_DONE){
+		lua_pushnil(L);
+		lua_pushliteral(L,"error sending message");
+		return 2;
+	}
+	lua_pushinteger(L,sent);
+	return 1;
+}
+
+/*-------------------------------------------------------------------------*\
+* Receives data from a netlink socket
+\*-------------------------------------------------------------------------*/
+static int meth_receive(lua_State *L){
+	p_netlink nl= (p_netlink) auxiliar_checkclass(L,"netlink{connected}",1);
+	struct nlmsghdr *nlh;
+	size_t got;
+	p_timeout tm = &nl->tm;
+
+	nlh=(struct nlmsghdr *)malloc(NLMSG_SPACE(MAX_PAYLOAD));
+	memset(nlh,0,NLMSG_SPACE(MAX_PAYLOAD));
+	nlh->nlmsg_len=NLMSG_SPACE(MAX_PAYLOAD);
+
+	int err= socket_recv(&nl->fd,(char *)nlh,nlh->nlmsg_len,&got,tm);
+    	if (err != IO_DONE && err != IO_CLOSED) {
+        	lua_pushnil(L);
+       		lua_pushstring(L,socket_strerror(err));
+        	return 2;
+    	}
+    	lua_pushinteger(L,got);
+    	lua_pushlstring(L,NLMSG_DATA(nlh), got);
+    	return 2;
+}
+
+/*-------------------------------------------------------------------------*\
+* Receives data and sender from a netlink socket
+\*-------------------------------------------------------------------------*/
+static int meth_receivefrom(lua_State *L) {
+	p_netlink nl= (p_netlink) auxiliar_checkclass(L,"netlink{unconnected}",1);
+	struct sockaddr_nl dst;
+	struct nlmsghdr *nlh;
+	size_t got;
+	p_timeout tm = &nl->tm;
+
+	nlh=(struct nlmsghdr *)malloc(NLMSG_SPACE(MAX_PAYLOAD));
+	memset(nlh,0,NLMSG_SPACE(MAX_PAYLOAD));
+	nlh->nlmsg_len=NLMSG_SPACE(MAX_PAYLOAD);
+
+	socklen_t len=sizeof(dst);
+	int err= socket_recvfrom(&nl->fd,(char *)nlh,nlh->nlmsg_len,&got,(struct sockaddr *)&dst,&len,tm);
+    	if (err != IO_DONE && err != IO_CLOSED) {
+        	lua_pushnil(L);
+       		lua_pushstring(L,socket_strerror(err));
+        	return 2;
+    	}
+    	lua_pushinteger(L,got);
+    	lua_pushlstring(L,NLMSG_DATA(nlh), got);
+	lua_pushinteger(L,nlh->nlmsg_pid);
+    	return 3;
+}
+
+/*-------------------------------------------------------------------------*\
+* Select support methods
+\*-------------------------------------------------------------------------*/
+static int meth_getfd(lua_State *L) {
+    p_netlink nl = (p_netlink) auxiliar_checkgroup(L, "netlink{any}", 1);
+    lua_pushnumber(L, (int) nl->fd);
+    return 1;
+}
+
+/* this is very dangerous, but can be handy for those that are brave enough */
```

Comment:
Is this is dangerous, why are we adding it here?

Reply:
Actually it is dangerous, but it is also handy in some situations. For example, we can use this to create make a socket in Lua to refer a socket created by 'libwebsockets' for example. Since LuaSocket does not provide a C API to do this, this sort of fallback becames handy.

We should add this anyway just to be compliant with what LuaSocket does currently, since we plan to contribute this code back to LuaSocket official repository.